daemons/fenced/fenced_history.c

```diff
@@ -271,40 +271,59 @@ stonith_xml_history_to_list(xmlNode *history)
  * \return The fence-history as xml
  */
 static xmlNode *
-stonith_local_history_diff(GHashTable *remote_history,
+stonith_local_history_diff_and_merge(GHashTable *remote_history,
                            gboolean add_id,
                            const char *target)
 {
     xmlNode *history = NULL;
+    GHashTableIter iter;
+    remote_fencing_op_t *op = NULL;
+    gboolean updated = FALSE;
     int cnt = 0;
 
     if (stonith_remote_op_list) {
-            GHashTableIter iter;
-            remote_fencing_op_t *op = NULL;
+            const char *key = NULL;
 
             history = create_xml_node(NULL, F_STONITH_HISTORY_LIST);
 
             g_hash_table_iter_init(&iter, stonith_remote_op_list);
-            while (g_hash_table_iter_next(&iter, NULL, (void **)&op)) {
+            while (g_hash_table_iter_next(&iter, (void **)&key, (void **)&op)) {
                 xmlNode *entry = NULL;
 
                 if (remote_history) {
                     remote_fencing_op_t *remote_op =
                         g_hash_table_lookup(remote_history, op->id);
 
                     if (remote_op) {
-                        if (remote_op->state != st_failed
-                            && remote_op->state != st_done
-                            && (op->state == st_failed
-                                || op->state == st_done)) {
+                        if (stonith_op_state_pending(op->state)
+                            && !stonith_op_state_pending(remote_op->state)) {
+
+                            crm_debug("Updating outdated pending operation %.8s "
+                                      "(state=%s) according to the one (state=%s) from "
+                                      "remote peer history",
+                                      op->id, stonith_op_state_str(op->state),
+                                      stonith_op_state_str(remote_op->state));
+
+                            g_hash_table_steal(remote_history, op->id);
+                            key = remote_op->id;
```

Comment:
Nobody is using key - right? Don't think it works as expected.
Would have gone for something like:

op->id = remote_op->id;
remote_op->id = key; /* why don't we call it id? */

Then the iter_replace blow will free op (without the id used in iter) together with the id previously in remote_op.

Reply:
I don't see exchanging the ids inside of the structures would bring anything. AFAICT, the `key` is still pointing to the old `op->id` which will be freed soon.  And according to the man page,  `g_hash_table_iter_replace(&iter, remote_op);` won't touch the `key`. I think we have to update `key` ourselves.

Reply:
Guess using iter_replace keeps the internal shuffling around of the hash-table at a minimum.
And that just doesn't allow exchanging the key. And I wouldn't dare doing more to a hash-table
while iterating through it than exchange the value behind a key.
As we don't register a separate free for the keys but have the key as part of the value and thus
free it together with the value things are unfortunately getting a bit clumsy.

Reply:
Indeed it's trickier than I thought ...